Add Deako integration

Balake opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->
## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
Add a Deako integration using the pydeako library.

### Deako devices
Deako devices are smart home light switches that communicate over BLE. Typical installations range from 1 device to over 50.

There is one device that is always online, as it's acting as the remote "bridge" from the internet to other devices on the network. All other devices get online periodically to check in and get potential updates. All devices currently reboot daily.  Since communication is over BLE, not all devices need to be on the network to control all devices, just one.

### Pydeako
The pydeako library has two main components: discovery and connection. pydeako serves as an integration to Deako devices, allowing for control of devices (power, dim, etc) over the network locally.

**Discovery** takes in a zeroconf instance and collects all the current addresses being advertised by Deako devices that are currently on the local network. This discovery client will continue to update this pool of addresses as devices come on the network or fall off.

**Connection** happens over a socket and the pydeako library maintains this connection. If the connection is dropped, a new address is retrieved and the process starts over. If a device is connected to over a socket, the device will remain online as long as it can. The device list and controls are implementations of the [Deako local API](https://github.com/DeakoLights/local-integrations/blob/master/API.md).

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/33554

Note: I had a previous PR which got closed because I didn't have time to address comments: https://github.com/home-assistant/core/pull/102190

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

Balake committed:
Deako integration using pydeako

joostlek reviewed (changes requested):
If we are using zeroconf, can't we use the built in zeroconf discovery method instead? It would then be 1 config entry = 1 device, and I think that's much more user friendly than having 1 config entry = n devices it can find on the network. It also allows for manual setup, which is something that is a big issue for integrations which use this discovery flow. So I'd rather see this use the normal zeroconf methods.

Balake commented:
> If we are using zeroconf, can't we use the built in zeroconf discovery method instead? It would then be 1 config entry = 1 device, and I think that's much more user friendly than having 1 config entry = n devices it can find on the network. It also allows for manual setup, which is something that is a big issue for integrations which use this discovery flow. So I'd rather see this use the normal zeroconf methods.

Thanks for the feedback!

What is the builtin zeroconf method? I didn't see it in the [docs](https://developers.home-assistant.io/docs/network_discovery/).

The Deako local integration is more like a bridge. So once you connect to one device, you then have access to all the devices since the primary communication between devices is BLE. The bridge device sends out control commands over BLE. Not all devices are on the network at once. We have some installations with over 50 devices, so that many devices on the network at once isn't feasible. They do check in periodically however. There's one device guaranteed to be online when users have setup network in the Deako app.

I hope I'm correctly understanding what you're suggesting.

joostlek commented:
https://developers.home-assistant.io/docs/config_entries_config_flow_handler#defining-steps
https://developers.home-assistant.io/docs/creating_integration_manifest#zeroconf

frenck commented:
I've marked this PR, as changes are requested that need to be processed.
Please un-draft it once it is ready for review again by clicking the **"Ready for review"** button.

Thanks! 👍 

../Frenck

[_Learn more about our pull request process._](https://developers.home-assistant.io/docs/review-process#prs-are-being-drafted-when-changes-are-needed)

Balake commented:
> https://developers.home-assistant.io/docs/config_entries_config_flow_handler#defining-steps https://developers.home-assistant.io/docs/creating_integration_manifest#zeroconf

Great thanks @joostlek. Just to be clear, the number of registered devices is not the number of devices that are found on the network, but rather the number of devices the device we connect to knows about through the Deako profile that's stored locally after being fetched from the Deako backend. I should have provided more context. The device we connect to is then the "bridge" over BLE to the other devices. 

The other main reason that I am passing zeroconf to the lib is that there is an address pool that is used in case of connections being dropped, devices rebooting, etc. I'll add some more context to the PR with how the devices work and how the library works.

joostlek commented:
But then still we like this method with the zeroconf config flow. Because it will actively tell the user that the device is on the network and ready to set up. This is an amazing user experience, so if we can use it, we would do that

Balake commented:
> But then still we like this method with the zeroconf config flow. Because it will actively tell the user that the device is on the network and ready to set up. This is an amazing user experience, so if we can use it, we would do that

Okay great, thanks for clarifying. I'll get these changes in.

Balake commented:
@joostlek I've added the config flow for this integration, adding both zeroconf discovery step and manual address input.

joostlek commented on homeassistant/components/deako/config_flow.py:
```diff
@@ -0,0 +1,126 @@
+        self, discovery_info: zeroconf.ZeroconfServiceInfo
+    ) -> ConfigFlowResult:
+        """Handle zeroconf discovery."""
+        # only one instance of deako integration should exist, device is a bridge
+        if self.hass.data.get(DOMAIN) is not None:
+            return self.async_abort(reason="already_configured")
+
+        await self.async_set_unique_id(DOMAIN)
```
Why do we enforce this? Why don't we allow if someone uses 2 deako bridges?

Balake replied:
Good question. We only need one device for a successful integration. A second bridge won't be beneficial or net any other devices. 

I tested removing this constraint and on larger installations, the discovered devices menu is flooded with all the devices that could be the bridge.

There's one instance I can see for two bridges and that would be separate installations on the same network. So you would have two sets of devices, but there's no way to distinguish what devices are a part of what installation from the advertisement data and even with the local API, there's no way to determine if a device is part of installation A or installation B without comparing the device list of multiple devices.

joostlek replied:
Can you elaborate on this? 
> the discovered devices menu is flooded with all the devices that could be the bridge.

Balake replied:
Yeah so I have access to a larger installation of Deako devices (>100) and due to their behavior with getting online to check in, there are always more than 10 online at a time roughly. 

By removing the constraint that there can only be one active config flow via the DOMAIN, the discovered section in the integrations menu will now have all the devices listed as separate Deako integrations. We only need one to act as the bridge, so having all those listed there is misleading.

joostlek replied:
And why should we connect to a bridge if we can also talk to the device directly?

Balake replied:
Not every device is guaranteed to be on the network and often most devices are not. The bridge device gives us two things:

1. the list of all known devices, whether on the network or not
2. access to the encrypted BLE mesh network for control requests

joostlek replied:
How can we detect if the device is a bridge?

Balake replied:
Good question! Any device that's on the network _can_ be the bridge device. The device that _is_ selected as the bridge device (aka connected to with a socket) will no longer advertise I believe.  I would have to confirm with the firmware team.

joostlek replied:
No but how do you know if the zeroconf message is the one sent by the bridge?

Balake replied:
All of the zeroconf advertisements are sent from devices that aren't the bridge, just ones that could be the bridge. This flow is essentially bridge selection where the first device that advertises is the bridge.

joostlek replied:
Mind sharing the zeroconf debug logs? Like I think I start to know what you're getting at, but I want to be sure.

Balake replied:
So I talked to firmware team, device continues to advertise when it's connected to as the bridge, but I think that's fine for the integration's use cases.

I'll get you some zeroconf logs on the larger installation that I have access to.

Balake replied:
Here are the logs:

[2024-07-17 11:06:18.log](https://github.com/user-attachments/files/16272949/2024-07-17.11.06.18.log)

joostlek replied:
Can you do `homeassistant.components.zeroconf: debug` instead?

Balake replied:
You bet! Much easier to read. 

[2024-07-17 12:51:08.log](https://github.com/user-attachments/files/16274057/2024-07-17.12.51.08.log)

joostlek commented:
Okay so I asked around because I still wasn't unsure about the zeroconf and I think I can now draw a complete picture

Option A: We store the IP of the bridge and we use the classic zeroconf flow. The biggest benefit is the quick and easy user experience with that the device can be discovered without user intervention.

Option B: We use the classic discovery flow. Which only discovers devices once you initiate the flow.

I double checked as I wasn't sure, and I didn't want to force a way of working that didn't work.

Balake commented:
> I double checked as I wasn't sure, and I didn't want to force a way of working that didn't work.

Option A: this is similar to what I have now, just not storing the IP of the bridge device as that can change with device reboots, but the pydeako library handles these cases since it utilizes a zeroconf service browser to maintain a queue of valid IPs in case a new bridge is needed.

Is option B this here? https://developers.home-assistant.io/docs/config_entries_config_flow_handler/#discoverable-integrations-that-require-no-authentication

Thanks for all the discussion. This is my first contribution to HA, so I want to make sure I'm doing things right and what I'm doing is understood.

joostlek commented:
> just not storing the IP of the bridge device as that can change with device reboot

HA handles would handle that then. The config entry gets a unique id, and the next time a zeroconf flow is started by a new mDNS packet, it will update the host of the entry with the same unique id

Option B would be what you had before.

Balake commented:
> > just not storing the IP of the bridge device as that can change with device reboot
> 
> HA handles would handle that then. The config entry gets a unique id, and the next time a zeroconf flow is started by a new mDNS packet, it will update the host of the entry with the same unique id
> 
> Option B would be what you had before.

Okay I see what you're saying. So how would that work when we have a working bridge device, a different device gets online and advertises over mDNS? Would our config entry just get overwritten with the new IP even though we had a good IP already?  Is there a way we can reject the update?

joostlek commented:
Reolink has something for this i noticed today. They check if the entity already exists and if it does they check if the connection is lost. If so, we apply the IP update, otherwise we do not. So yes, its possible, but if this is not breaking id recommend this logic for s followup

Balake commented:
> Reolink has something for this i noticed today. They check if the entity already exists and if it does they check if the connection is lost. If so, we apply the IP update, otherwise we do not. So yes, its possible, but if this is not breaking id recommend this logic for s followup

Okay thanks for the example! I think that might be a good long term goal, but would require quite invasive changes to the lib I'm using (something I can get in at some point though). 

For the time being, I think maybe the classic discovery flow might be our best bet. I suspect the pattern with my changes currently isn't very common. With my testing, auto-discovery works well. If you're okay with that, I'll go ahead and get those changes back in and we can go from there.

joostlek commented:
Sure!

Balake commented:
> Sure!

Sweet, will get those changes in. Thanks again!

Balake commented:
Alright @joostlek, back to classic discovery flow! Let me know if there's anything else I can explain or add to.

joostlek commented on homeassistant/components/deako/__init__.py:
```diff
@@ -0,0 +1,62 @@
+        await connection.disconnect()
+        raise ConfigEntryNotReady(devices)
+
+    hass.data.setdefault(DOMAIN, {})[entry.entry_id] = connection
```
Let's use the `entry.runtime_data`, there's a dev blog post about that a month or 2 ago, I'm on mobile so I dont have a link

Balake replied:
Ooooo I like that. Thanks! Certainly cleans things up a bit.

joostlek commented on homeassistant/components/deako/light.py:
```diff
@@ -0,0 +1,102 @@
+    client: Deako = hass.data[DOMAIN][config.entry_id]
+
+    devices = client.get_devices()
+    lights = [DeakoLightEntity(client, uuid) for uuid in devices]
+    add_entities(lights)
```
Can be merged

joostlek commented on homeassistant/components/deako/light.py:
```diff
@@ -0,0 +1,102 @@
+class DeakoLightEntity(LightEntity):
+    """Deako LightEntity class."""
+
+    # retype because these will be set
+    _attr_unique_id: str
```
Let's not retype unique id

Balake replied:
Got it, I'll throw in some asserts

joostlek commented on homeassistant/components/deako/light.py:
```diff
@@ -0,0 +1,102 @@
+        self.update()
+        self.schedule_update_ha_state()
+
+    def get_state(self) -> dict:
```
Can we type the dict better?

joostlek commented on homeassistant/components/deako/manifest.json:
```diff
@@ -0,0 +1,12 @@
+{
```
Let's also add `single_config_entry: true`

joostlek commented on tests/components/deako/conftest.py:
```diff
@@ -0,0 +1,38 @@
+    return MockConfigEntry(
+        domain=DOMAIN,
+        data={},
+        unique_id="aabbccddeeff",
```
You don't have an unique id

Balake replied:
Good catch!

joostlek commented on tests/components/deako/conftest.py:
```diff
@@ -0,0 +1,38 @@
+    )
+
+
+@pytest.fixture(autouse=True)
+def deako_mock_async_zeroconf(mock_async_zeroconf: MagicMock):
+    """Auto mock zeroconf."""
```
What's this for?

Balake replied:
Looks like I never used it. Not needed anyway. Thanks!

joostlek commented on tests/components/deako/test_config_flow.py:
```diff
@@ -0,0 +1,45 @@
+from homeassistant.core import HomeAssistant
+
+
+@pytest.mark.asyncio
+async def test_deako_async_has_devices(
+    hass: HomeAssistant,
+    mock_async_zeroconf: None,
+) -> None:
+    """Test successful device discovery."""
+
+    with patch(
+        "homeassistant.components.deako.config_flow.DeakoDiscoverer", autospec=True
+    ) as mock_discoverer:
+        ret = await _async_has_devices(hass)
+
+        assert ret
+        mock_discoverer.assert_called_once()
+        mock_discoverer.return_value.get_address.assert_called_once()
+
+
+@pytest.mark.asyncio
+async def test_deako_async_has_devices_error(
+    hass: HomeAssistant,
+    mock_async_zeroconf: None,
+) -> None:
+    """Test successful device discovery."""
+
+    with patch(
+        "homeassistant.components.deako.config_flow.DeakoDiscoverer", autospec=True
+    ) as mock_discoverer:
+        mock_discoverer.return_value.get_address.side_effect = (
+            DevicesNotFoundException()
+        )
+        ret = await _async_has_devices(hass)
+
+        assert not ret
+        mock_discoverer.assert_called_once()
+        mock_discoverer.return_value.get_address.assert_called_once()
```
The tests should be more end to end. Aka, we should patch the library appropriately, start a config flow, and then check the results

Balake replied:
I gotcha, that makes sense, thanks!

joostlek commented on tests/components/deako/test_init.py:
```diff
@@ -0,0 +1,127 @@
+
+    mock_config_entry.add_to_hass(hass)
+
+    with patch.object(
+        hass.config_entries, "async_forward_entry_setups"
+    ) as async_forward_entry_setups_mock:
```
We should not patch this

joostlek commented on tests/components/deako/test_init.py:
```diff
@@ -0,0 +1,127 @@
+    assert mock_config_entry.state is ConfigEntryState.SETUP_RETRY
+
+
+@pytest.mark.asyncio
```
Not needed

joostlek commented on tests/components/deako/test_light.py:
```diff
@@ -0,0 +1,259 @@
+    pydeako_deako_mock: MagicMock,
+    mock_config_entry: MockConfigEntry,
+) -> None:
+    """Test light platform setup with no devices returned."""
```
This test is strange because this will never happen in real life because init has a check for it

Balake replied:
Yeah you're right. Was more of an isolated test, not necessarily practical

joostlek commented on tests/components/deako/test_light.py:
```diff
@@ -0,0 +1,259 @@
+    }
+    pydeako_deako_mock.return_value.is_dimmable.return_value = False
+
+    await hass.config_entries.async_setup(mock_config_entry.entry_id)
+    await hass.async_block_till_done()
+
+    # make sure our initial state exists
+    assert (state := hass.states.get(f"{LIGHT_DOMAIN}.kitchen"))
+    assert state.state == STATE_OFF
+
+    # test other properties
+    entity = entity_registry.async_get(f"{LIGHT_DOMAIN}.kitchen")
+    assert entity
+    assert entity.unique_id == "uuid"
+    assert entity.original_name is None
+    assert ColorMode.ONOFF in entity.capabilities.get("supported_color_modes")
```
You can look into snapshotting, use snapshot_platform to make sure you snapshot the entity and the state

Balake committed:
fix: address feedback

- make unit tests more e2e
- use runtime_data to store connection

Balake commented:
@joostlek I believe I addressed your comments

joostlek commented:
Haven't found time yet, will do so in the next few days

joostlek reviewed (changes requested):
Oh wait, apparently I already had some comments in draft. Not sure if I completed this round but here's something to look at for the time being

Balake commented:
@joostlek thanks for all the feedback. I'm learning a lot! I'll get going on some of these changes.

Balake commented on homeassistant/components/deako/light.py:
```diff
@@ -0,0 +1,101 @@
+class DeakoLightEntity(LightEntity):
+    """Deako LightEntity class."""
+
+    _attr_supported_color_modes: set[ColorMode]
```
This was so that I didn't have to do a type check to use the `in` operator because I knew it was going to be a set, but I think this was a mistake. No need to overwrite the type, good call out.

Balake commented on tests/components/deako/conftest.py:
```diff
@@ -0,0 +1,31 @@
+
+
+@pytest.fixture(name="pydeako_discoverer_mock", autouse=True)
+def pydeako_discoverer_mock(mock_async_zeroconf: MagicMock):
```
This is needed since it's a dep to the integration and the deako discoverer takes in a zeroconf instance and then subscribes to mDNS additions/updates/removals. Without this, I get this for almost all my tests:

`Unable to set up dependencies of 'deako'. Setup failed for dependencies: ['zeroconf']`

If there's a better way I should do this, I'm open to suggestions.

Balake commented on tests/components/deako/conftest.py:
```diff
@@ -0,0 +1,31 @@
+    )
+
+
+@pytest.fixture(name="pydeako_deako_mock", autouse=True)
+def pydeako_deako_mock():
+    """Mock pydeako deako client."""
+    with patch("homeassistant.components.deako.Deako", autospec=True) as mock:
+        yield mock
+
+
+@pytest.fixture(name="pydeako_discoverer_mock", autouse=True)
+def pydeako_discoverer_mock(mock_async_zeroconf: MagicMock):
+    """Mock pydeako discovery client."""
+    with patch("homeassistant.components.deako.DeakoDiscoverer", autospec=True) as mock:
+        yield mock
```
Oooooo I good suggestion, thanks!

Balake commented on tests/components/deako/test_config_flow.py:
```diff
@@ -0,0 +1,83 @@
+
+        assert result.get("type") is FlowResultType.ABORT
+        assert result.get("reason") == "single_instance_allowed"
+        mock_discoverer.return_value.get_address.assert_not_called()
```
I'll go ahead and remove it as it's something that is provided by HA in the manifest, so if that doesn't work, we will have bigger issues.

Balake commented on homeassistant/components/deako/manifest.json:
```diff
@@ -0,0 +1,13 @@
+  "loggers": ["pydeako"],
+  "requirements": ["pydeako==0.4.0"],
+  "single_config_entry": true,
+  "zeroconf": ["_deako._tcp.local."]
```
Yes it does!

![Screenshot 2024-08-12 at 12 57 38 PM](https://github.com/user-attachments/assets/3ebf2eca-8b15-4c1c-a5bc-dac07b1da44c)

Balake commented on tests/components/deako/test_config_flow.py:
Good suggestion.

Balake committed:
fix: address feedback part 2

- added better type safety for Deako config entries
- refactored the config flow tests to use a conftest mock instead of directly patching
- removed pytest.mark.asyncio test decorators

joostlek commented on homeassistant/components/deako/__init__.py:
```diff
@@ -0,0 +1,64 @@
+PLATFORMS: list[Platform] = [Platform.LIGHT]
+
+
+class DeakoConfigEntry(ConfigEntry):
+    """Typed config entry."""
+
+    runtime_data: Deako | None
```
```suggestion
type DeakoConfigEntry = ConfigEntry[Deako]
```

Balake replied:
Oh sweet, thanks!

joostlek commented on homeassistant/components/deako/__init__.py:
```diff
@@ -0,0 +1,64 @@
+
+async def async_unload_entry(hass: HomeAssistant, entry: DeakoConfigEntry) -> bool:
+    """Unload a config entry."""
+    if entry.runtime_data is not None:
+        await entry.runtime_data.disconnect()
```
```suggestion
    await entry.runtime_data.disconnect()
```

joostlek commented on homeassistant/components/deako/light.py:
```diff
@@ -0,0 +1,106 @@
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config: DeakoConfigEntry,
+    add_entities: AddEntitiesCallback,
+) -> None:
+    """Configure the platform."""
+    client = config.runtime_data
+
+    if client is not None:
+        add_entities([DeakoLightEntity(client, uuid) for uuid in client.get_devices()])
+    else:
+        _LOGGER.error("Deako client not set in config entry")
```
```suggestion
    async_add_entities: AddEntitiesCallback,
) -> None:
    """Configure the platform."""
    client = config.runtime_data

    async_add_entities([DeakoLightEntity(client, uuid) for uuid in client.get_devices()])
```

joostlek commented on homeassistant/components/deako/light.py:
```diff
@@ -0,0 +1,106 @@
+
+    def update(self) -> None:
+        """Call to update state."""
+        state = self.get_state()
```
why is this a separate function?

Balake replied:
Good question.... I'm not sure I have a good reason, I'll consolidate.

joostlek commented on tests/components/deako/test_config_flow.py:
Ideally we also have a test that adds the mock_config_entry to hass, and then tries to start a new flow. It should abort because we only allow one config entry

Balake replied:
Got it. I misunderstood one of you previous comments and removed this test.

joostlek reviewed (changes requested):
Last bits

Balake committed:
fix: address feedback pt 3

- simplify config entry type
- add test for single_instance_allowed
- remove light.py get_state(), only used once, no need to be separate function

joostlek reviewed (changes requested):
The CI is failing, can you take a look?

Balake commented:
> The CI is failing, can you take a look?

Sure thing!

Balake committed:
fix: ruff format

Balake commented:
@joostlek I fixed CI, is there anything else that I'm missing?

joostlek commented:
Oh, wanted to apply it but I don't have the rights

Balake committed:
Update homeassistant/components/deako/__init__.py

Co-authored-by: Joost Lekkerkerker <joostlek@outlook.com>

joostlek commented:
Let me kick off the CI.
Will probably merge tomorrow as I'm now trying to survive the wind at the Dutch grand Prix

Balake commented:
> Let me kick off the CI. Will probably merge tomorrow as I'm now trying to survive the wind at the Dutch grand Prix

Sweet, thanks! Good luck over there 🤞

Balake commented:
@joostlek did you survive?

joostlek commented:
Shit. Forgot to merge and beta just has been cut. I already had the idea I forgot something.

Balake commented:
> Shit. Forgot to merge and beta just has been cut. I already had the idea I forgot something.

No worries! Thanks for all the help and the review!
